### ISSUE TITLE
Proposal for a New `symfony/access-control` Component to Decouple Authorization

### ISSUE THREAD

**Issue by Spomky** [symfony/symfony#59300]
### Description

Currently, the `Security` component mixes **authentication** and **authorization** together. It also assumes a **Role-Based Access Control** (RBAC) model, forcing `UserInterface::getRoles()`. However, many applications:

- Do **not** need roles at all (e.g., policy-based, attribute-based, or ownership checks),
- Only require authorization (e.g., CRON jobs, system tasks) but **not** full user authentication,
- Need alternative models such as **Discretionary Access Control (DAC)**, **Mandatory Access Control (MAC)**, **Attribute-Based Access Control (ABAC)**, or **Risk-Based** decision-making,
- Or simply handle **authorization** for subjects other than a typical user (e.g., microservices, batch processes, system-level services).

---

### Proposed Change

- **Create a `symfony/access-control` Component**  
  - Migrate **all authorization** features from `symfony/security` (voters, `is_granted()` Twig functions, role checks, etc.) into a dedicated package.  
  - This new component can offer built-in **RBAC**, **ABAC**, **DAC**, **MAC**, **Risk-Based** policies, or custom rule sets—without forcing roles.  
  - **Subject-Agnostic**: It should not rely on having a `UserInterface` at all, allowing any entity or context to be the “subject” of an authorization check.

- **Deprecate `getRoles()` in `UserInterface`**  
  - Let developers define their own permission models or attributes. Roles can still be supported, but not required.

- **Keep `symfony/security` Focused on Authentication**  
  - Firewalls, token management, and user identification remain in `symfony/security`.  
  - `symfony/access-control` handles how to grant or deny actions, regardless of authentication approach or subject type.

---

### Potential Benefits

1. **Cleaner Separation of Concerns**  
   - Authentication/identification in one place, advanced authorization in another.

2. **Flexible Access Control Models**  
   - Developers can choose RBAC, ABAC, DAC, MAC, risk-based checks, or any custom logic without coupling to roles.

3. **Scalable for Complex Scenarios**  
   - Systems can implement stricter or more dynamic policies without hacking around `UserInterface::getRoles()`.

4. **Works for Authorization-Only Apps**  
   - Service-to-service APIs, batch processes, or microservices needing permission checks can use the access control piece alone.

5. **Subject-Agnostic Design**  
   - Easily handle use cases where the “actor” is not a typical user but any other subject (e.g., an automated system or a machine identity).

---

### Conclusion

By introducing a dedicated `symfony/access-control` component and separating it from `symfony/security`, Symfony would better accommodate applications that either:

- Need robust authorization **without** roles,
- Rely on alternative models (e.g., ABAC, DAC, MAC, Risk-Based),
- Only require **authorization** without full-fledged **authentication**, or
- Wish to treat any entity as the “subject” of access checks, not just a `UserInterface`.

Feedback on backward compatibility, migration strategy, and community support is greatly appreciated!

**Comment by Spomky** [comment]
CC @chalasr, @wouterj, @welcoMattic

**Comment by chalasr** [comment]
I support the goal here. It’s probably worth mentioning that the security component is actually composed of 2 main ones that are security-core and security-http, which both suffer from the coupling we are talking about. 

Providing guidance around BC is limited at this stage - we would need to see more precisely what APIs we want before tempering it with detailed BC concerns. 
All I can say is we should strive to limit refactoring to the minimum viable that make it possible to reach the goals, no big bang. Also worth reminding that we don’t break stuff suddenly in major versions but always provide smooth upgrade paths :)

**Comment by Spomky** [comment]
Indeed, I’m not planning to break everything at once or make a “big bang” change.
Instead, my idea is to create a dedicated component that preserves the existing functionalities, but with a new design. This allows us to maintain backward compatibility while moving forward. Once we’re confident in the stability and maturity of the new component, we can gradually deprecate the old features—likely in the 8.x or 9.x versions—without disrupting existing developers.
This incremental approach should ensure a smooth migration path, reflecting our commitment to balanced evolution rather than sudden, large-scale changes.

**Comment by wouterj** [comment]
Providing other authorization mechanism than role-based in Symfony sounds very interesting. That would allow us to remove the last 'unrelated' method from the `UserInterface`.

However, I'm not seeing much advantages of separating this into a new component, rather than adding this in the existing component. My feeling now is that adding more authorization features one by one to the Security component will be much easier to review and introduce to the community (both "marketing-wise" as code/compatibility-wise). But maybe I'm missing relevant context in this regard? (currently, the only benefit I see is downloading less code on install when you only need authorization in your full app)

**Comment by Neirda24** [comment]
I think decoupling authorization and authentication is great ! Authorization should not require authentication to be installed !

**Comment by gggeek** [comment]
My 2c: having recently looked at the auth/authz code found in Symfony (version 6.4, must admit) in order not to use it as-is, but rather as inspiration for creating a bespoke auth system for a custom app (which has good reasons for not using composer nor external code), I must say that, while I find it well architected and flexible enough to cater to almost any requirement, it is somewhat daunting in complexity. Esp. the split of classes between security-core and security-http made my brain ache.

This is a long-winded way of saying that, while I applaud the thinking of this proposal, I am afraid that yet more complexity will creep in in the end...

**Comment by Spomky** [comment]
> Providing other authorization mechanism than role-based in Symfony sounds very interesting. That would allow us to remove the last 'unrelated' method from the UserInterface.

Hello @wouterj,
I was not aware of https://github.com/symfony/symfony/pull/59106 and indeed it we lead to a very simple `UserInterface` with a unique method `getIdentifier()`.

> My feeling now is that adding more authorization features one by one to the Security component will be much easier to review and introduce to the community (both "marketing-wise" as code/compatibility-wise).

That was my initial impression as well. The issue is that roles are heavily integrated into the current security components—not just in the UserInterface, but also in the security tokens and configuration. Introducing new access control classes, interfaces, or enums will require a careful handling of deprecations that we’ll support through Symfony 9.4, and I’m concerned that this could be quite challenging.

As I explained in my initial post, the security component should not mix authentication and authorization. That’s why I propose to start a new approach while keeping the existing system intact, allowing users to opt in to the newer features.

> This is a long-winded way of saying that, while I applaud the thinking of this proposal, I am afraid that yet more complexity will creep in in the end...

@gggeek I’ve started working on this new component and will share it in the coming days.
There’s nothing fundamentally new, so users won’t be confused by this component. The main idea is to build on existing features—like Role-based Access Control—using similar attribute and service. Voters will still cast their votes on a (attribute, subject) tuple, depending on the context (user, system, etc.). The  strategies will then make a decision based on those votes.

**Comment by stof** [comment]
The authorization system of the security component is not restricted to roles. The `AuthorizationCheckerInterface` is granting access on some attributes, which are **not** roles. They are whatever the registered voters support (our RoleVoter will vote on attributes being strings starting with `ROLE_`, comparing them to the roles provided in the token being the result of the authentication, but that's the only usage of roles in the authorization system in core).

And this system already has a concept of `subject` for the check, which is **not** the UserInterface. I'm not sure whether your usage of `subject` refers to the same or a different concept.
And note that this UserInterface representing the actor does not actually enforce that actor to be a human. A service-to-service use case would still require that other service to authenticate and to be represented in your system, which is currently done by implementing UserInterface in a different way.
I don't see how we could have a new access-control component that is fully decoupled from the authentication component, as the voter will have to vote on checks for a specific actor (and making voters read it from the TokenStorage instead of receiving it as argument would make it impossible to migrate the `isGrantedForUser` feature, so that would be a regression to me, even without considering the migration cost).
We already have separate parts of the security component responsible for authorization and authentication. If the new component has a dependency on the security component due to the TokenInterface in the signature, it only moves classes around with no benefits to outweight the migration cost.

**Comment by Spomky** [comment]
Hello @stof,

Thank you for your feedback!
I’ve created the PR #59439 so we can discuss these ideas in a more concrete way than in this Issue.

The code in the PR demonstrates the approach I’m proposing and should help clarify how the `subject` could be handled and how authentication and authorization might work together.
Please feel free to continue the discussion in the PR so we can gather your feedback and refine the proposal.

**Comment by Spomky** [comment]
Closing as explained in #59439
